src/main/java/build/buf/protovalidate/Violation.java

```diff
@@ -0,0 +1,131 @@
+ * violation.
+ */
+public class Violation {
+  private final build.buf.validate.Violation proto;
+  @Nullable Value fieldValue;
+  @Nullable Value ruleValue;
+
+  /** Builds a Violation instance. */
+  public static class Builder {
+    @Nullable private build.buf.validate.Violation proto;
+    @Nullable Value fieldValue;
+    @Nullable Value ruleValue;
+
+    /**
+     * Sets the underlying protobuf message that corresponds to the violation.
+     *
+     * @param proto The value to set for this field.
+     * @return The builder.
+     */
+    public Builder setProto(build.buf.validate.Violation proto) {
+      this.proto = proto;
+      return this;
+    }
+
+    /**
+     * Sets the field value that corresponds to the violation.
+     *
+     * @param fieldValue The field value corresponding to this violation.
+     * @return The builder.
+     */
+    public Builder setFieldValue(@Nullable Value fieldValue) {
+      this.fieldValue = fieldValue;
+      return this;
+    }
+
+    /**
+     * Sets the rule value that corresponds to the violation.
+     *
+     * @param ruleValue The rule value corresponding to this violation.
+     * @return The builder.
+     */
+    public Builder setRuleValue(@Nullable Value ruleValue) {
+      this.ruleValue = ruleValue;
+      return this;
+    }
+
+    /**
+     * Builds a Violation instance with the provided parameters.
+     *
+     * @return A Violation instance.
+     */
+    public Violation build() {
+      return new Violation(Objects.requireNonNull(proto), fieldValue, ruleValue);
+    }
+
+    private Builder() {}
+  }
+
+  /**
+   * Constructs a new empty builder.
+   *
+   * @return A new empty builder instance.
+   */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  Violation(
+      build.buf.validate.Violation proto, @Nullable Value fieldValue, @Nullable Value ruleValue) {
+    this.proto = proto;
+    this.fieldValue = fieldValue;
+    this.ruleValue = ruleValue;
+  }
+
+  /**
+   * Gets the protobuf data that corresponds to this constraint violation.
+   *
+   * @return The protobuf violation data.
+   */
+  public build.buf.validate.Violation getProto() {
+    return proto;
+  }
+
+  /**
+   * Gets the field value that corresponds to the violation.
+   *
+   * @return The field value corresponding to this violation.
+   */
+  public @Nullable Value getFieldValue() {
+    return fieldValue;
+  }
+
+  /**
+   * Gets the rule value that corresponds to the violation.
+   *
+   * @return The rule value corresponding to this violation.
+   */
+  public @Nullable Value getRuleValue() {
+    return ruleValue;
+  }
+
+  /**
+   * Constructs a {@link Builder} with all fields set to match this {@link Violation}.
+   *
+   * @return A new {@link Builder} instance.
+   */
+  public Builder toBuilder() {
+    return new Builder().setProto(proto).setFieldValue(fieldValue).setRuleValue(ruleValue);
+  }
```

Comment:
Is this useful? Seems unnecessary and unlikely to be used.

Reply:
Never mind. I see this was to make it easier to insert the new non-proto into existing code that already made use of `toBuilder` all over the place, to revise/mutate the protos as the result was popped-up the stack.

That is rather unfortunate. In addition to meaning extra public API that we don't expect users to actually find useful, it also means potentially _lots_ of re-building new objects. It probably should have popped builders all the way up the stack so it could just mutate and then `build()` only at the end when returning the value to the caller... Though a separate internal accumulator type would likely have been even better.

I suspect it might be a pain, but maybe worth tackling anyway? (Maybe not in this PR though, but in a follow-up, before we release this stuff, to minimize breaking changes?)

Reply:
What I really wanted to do was only expose an interface for the `Violation` and keep the actual implementation internal, but it felt like there was a lot to untangle to get it there.

That said, I like the idea of having an internal accumulator type. If I do make that separation throughout the entire codebase, then it's probably not even necessary to expose a concrete `Violation` class since nothing in protovalidate is going to accept a `Violation` object anyway. Or is that a bad idea?